test/test_tagging.py

```diff
@@ -0,0 +1,235 @@
+import json
+from typing import (
+    Optional,
+)
+import unittest
+from unittest.mock import (
+    MagicMock,
+    patch,
+)
+
+from azul.deployment import (
+    populate_tags,
+)
+from azul.logging import (
+    configure_test_logging,
+)
+from azul_test_case import (
+    AzulUnitTestCase,
+)
+
+
+def setupModule():
+    configure_test_logging()
+
+
+class TestTerraformResourceTags(AzulUnitTestCase):
+
+    def assertDictEqualPermissive(self, expected, actual):
+        """
+        Assert wrapper for permissive_compare.
+        """
+        path = self.permissive_compare(expected, actual)
+        return self.assertIsNone(path, f'Discrepancy in path: {path}')
+
+    def permissive_compare(self, expected, actual, *path: str) -> Optional[str]:
+        """
+        Recursive dictionary comparison. Skips comparing values when the
+        reference value is set to None.
+
+        Returns None if dictionaries match, or path of discrepancy as a
+        string otherwise.
+
+        >>> t = TestTerraformResourceTags()
```

Comment:
These doctests don't seem to cover all the branches:

![image](https://user-images.githubusercontent.com/5143256/94844161-aae3bf80-03d2-11eb-9175-fe30bd8ea0b3.png)